### PR TITLE
Easy epoch selection for dumping state parts and log the number of nodes visited

### PR DESCRIPTION
**Pull request by nikurt**
This lets us easily script a tool to dump the state of each epoch, without running extra commands to find the correct `--block-hash` value.

### CODE REVIEW

File: tools/state-viewer/src/dump_state_parts.rs [hunk 1]
```diff
@@ -1,15 +1,96 @@
...
-use near_primitives_core::types::ShardId;
+use near_primitives_core::types::{BlockHeight, EpochHeight, ShardId};
 use near_store::Store;
 use nearcore::{NearConfig, NightshadeRuntime};
 use std::path::Path;
+use std::str::FromStr;
 use std::sync::Arc;
 
+#[derive(Subcommand, Debug, Clone)]
+pub(crate) enum EpochSelection {
```
Comment (mm-near) [inline tools/state-viewer/src/dump_state_parts.rs:18]: nice.

File: tools/state-viewer/src/dump_state_parts.rs [hunk 2]
```diff
@@ -1,15 +1,96 @@
...
+    /// Fetch an epoch containing the given block hash.
+    BlockHash { block_hash: String },
+    /// Fetch an epoch containing the given block height.
+    BlockHeight { block_height: BlockHeight },
+}
+
+fn get_epoch_id(
+    epoch_selection: EpochSelection,
+    store: Store,
+    chain_store: &mut ChainStore,
```
Comment (mm-near) [inline tools/state-viewer/src/dump_state_parts.rs:34]: does this and epoch_manager have to be 'mut'? (seems strange..)

File: tools/state-viewer/src/dump_state_parts.rs [hunk 3]
```diff
@@ -1,15 +1,96 @@
...
+    EpochId { epoch_id: String },
+    /// Fetch epochs at the given height.
+    EpochHeight { epoch_height: EpochHeight },
+    /// Fetch an epoch containing the given block hash.
+    BlockHash { block_hash: String },
+    /// Fetch an epoch containing the given block height.
+    BlockHeight { block_height: BlockHeight },
+}
+
+fn get_epoch_id(
```
Comment (mm-near) [inline tools/state-viewer/src/dump_state_parts.rs:31]: nit:

I'd actually make this function as 'impl' function of EpochSelector:

impl EpochSelector {
pub to_epoch_id(store, chain_store, epoch_maanger) -> EpochId

File: tools/state-viewer/src/dump_state_parts.rs [hunk 4]
```diff
@@ -1,15 +1,96 @@
...
+        }
+        EpochSelection::BlockHeight { block_height } => {
+            // Fetch an epoch containing the given block height.
+            let block_hash = chain_store.get_block_hash_by_height(block_height).unwrap();
+            epoch_manager.get_epoch_id(&block_hash).unwrap()
+        }
+    }
+}
+
+fn get_prev_hash_of_epoch(
```
Comment (mm-near) [inline tools/state-viewer/src/dump_state_parts.rs:63]: comment

File: tools/state-viewer/src/dump_state_parts.rs [hunk 5]
```diff
@@ -1,15 +1,96 @@
...
+    epoch_info: &EpochInfo,
+    chain_store: &ChainStore,
+    epoch_manager: &mut EpochManager,
+) -> CryptoHash {
+    let head = chain_store.head().unwrap();
+    let mut cur_block_info = epoch_manager.get_block_info(&head.last_block_hash).unwrap();
+    loop {
```
Comment (mm-near) [inline tools/state-viewer/src/dump_state_parts.rs:70]: I'm surprised that you have to iterate over epochs like this. Could you just fetch the epoch from the epoch manager directly?

**Review by mm-near — CHANGES_REQUESTED** [review]

File: core/store/src/trie/iterator.rs [hunk 6]
```diff
@@ -350,6 +350,7 @@ impl<'a> TrieIterator<'a> {
                 }
             }
         }
+        tracing::debug!(target: "runtime", num_nodes = nodes_list.len());
```
[the remark this answers is not in the document]
Comment (nikurt) [reply core/store/src/trie/iterator.rs:353]: The context is in the span created above. Example:
```
2022-11-23T10:47:08.876915Z DEBUG obtain_state_part{part_id=1 shard_id=2 block_hash=7DmJMS8kjuPS4d7UcKTwSzxqb8aYLxuEPug4UP1RUPZ9 num_parts=57723}:visit_nodes_interval: runtime: num_nodes=11375
```

Moved it to the callsite.

File: tools/state-viewer/src/dump_state_parts.rs (hunk 5 above)
Comment (nikurt) [reply tools/state-viewer/src/dump_state_parts.rs:70]: Fetching an epoch is not enough. I need a block hash of the last block before the specified epoch.

I can fetch the epoch, but `EpochManager` has no functions mapping epochs to blocks.
`BlockInfo` has `epoch_first_block`, but I can't get any `BlockInfo` from `EpochInfo`.

Same is true for `RuntimeAdapter`. I could call `get_validator_info(EpochInfo)` to get `epoch_start_height`, but to get a block by height, I need an instance of `Chain`.

Comment (nikurt) [reply tools/state-viewer/src/dump_state_parts.rs:70]: Please let me know if I'm missing something obvious in the APIs.

Extending the API involves adding unit tests and faking the new methods. As that API is not needed for running the node, I prefer to workaround that.

Comment (mm-near) [reply tools/state-viewer/src/dump_state_parts.rs:82]: thanks for the explanation.

**Review by mm-near — APPROVED** [review]

**Pull request merged**